### PR TITLE
Refactor Version class to make version bumps easier

### PR DESCRIPTION
**Pull request by alpar-t**
With this change we only have to add one line to add a new version.
The intent is to make it less error prone and easier to write a script
to automate the process.

There are tools to generate source code or byte-code that would allow us to 
automate version bumps based on a plain list of versions,
but that would create friction with IDEs.
There are also tools that semantically understand java source and are able to reformat it, 
but unfortunately there's no existing tooling to edit a java source file. 
With this change it should be straight forward to automate it by manipulating the source 
as a plain text file.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/Version.java
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/Version.java]: Let's also:
 - only add for fields named V_\d+_\d+_\d+
 - assert the only field not matching this name is CURRENT
 - assert that the id is only added to the map once (currently it's not because V_8_0_0 and CURRENT)

File: server/src/main/java/org/elasticsearch/Version.java
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/Version.java]: Let's use a primitive map, `ImmutableOpenIntMap`.

**Review by jasontedor — CHANGES_REQUESTED** [review]
Looks good, I left some comments.

File: server/src/main/java/org/elasticsearch/Version.java
Comment (jasontedor) [reply server/src/main/java/org/elasticsearch/Version.java]: Also, let's assert that the id matches the value derived from the name, so that we can reduce the possibility of error even further!

**Comment by jasontedor** [comment]
@atorok I pushed a commit with my suggestions.

**Review by jasontedor — APPROVED** [review]
LGTM, but since I pushed commits I think that @rjernst should review too. 😇

**Review by rjernst — APPROVED** [review]
This is a long needed change, thanks! LGTM

**Comment by mark-vieira** [comment]
> There are tools to generate source code or byte-code that would allow us to
automate version bumps based on a plain list of versions,
but that would create friction with IDEs.

If we want the best tooling available, using an annotation processor would probably be the best bet. Modern IDEs have support for those. I'm generally not a fan of code generation, but if storing version info in a structured file and generating the constants makes version bumping less of a pain then I'm all for it.

**Comment by jasontedor** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by jasontedor** [comment]
> I'm generally not a fan of code generation, but if storing version info in a structured file and generating the constants makes version bumping less of a pain then I'm all for it.

I’m not a fan either. And anyway @rjernst and me have discussed how to do this across the stack. Code generation isn’t effective here because now we need generators for a bunch of different projects when most version bumps are capably done by `sed`. Only Elasticsearch is special but it can be done with text manipulation too, this PR makes it even easier.

**Comment by atorok** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by atorok** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by atorok** [comment]
I'm not sure what happened with the history here, I am rebasing on master to fix it without any additional changes

**Pull request merged**
